File: Java/Othello/src/othello/Pool.java
/**
 * @author Copyright (C) 2018, Christopher N. Hume. All rights reserved.
 *
 * You should have received a copy of the MIT License along with this program.
 * If not, see https://opensource.org/licenses/MIT.
 *
 * For background on the development of Reversi and Othello
 * see https://en.wikipedia.org/wiki/Reversi
 */
package othello;
import java.util.Stack;

class Pool<T> {
  final int nDefaultAllocation = 12;
  public int activeCount;
  public int activePeak;
  private Stack<T> inactive;

  public String name;
  private Factory<T> factory;

  public Pool(Factory<T> factory, String name) {
    this.factory = factory;
    this.name = name;
    inactive = new Stack<T>();
    Clear();
  }

  public void Clear() {
    activePeak = activeCount = 0;
  }

  private void Allocate(int nAllocations) {
    for (int n = 0; n < nAllocations; n++)
      inactive.push(factory.create());
  }

  public T push() {
    if (inactive.size() == 0)
      Allocate(nDefaultAllocation);

    T top = inactive.pop();
    assert top != null;			// Empty inactive Pool

    IncActive();
    return top;
  }

  public void pop(T top) throws NullPointerException {
    if (top == null)
      throw new NullPointerException("Pool.pop() called with null argument");

    inactive.push(top);
    DecActive();
  }

  private void IncActive() {
    if (activePeak < ++activeCount) {
      activePeak = activeCount;
      //[Debug Peak]DisplayActive();
    }
  }

  private void DecActive() {
    activeCount--;
  }

  public void DisplayActive() {
    Logger.logLevel(
      Logger.DEBUG,
      String.format("%s Count = %d, Peak = %d", name, activeCount, activePeak));
  }
}

File: CSharp/Square/Square/Resource/Pool/Pool.cs
//
// Copyright (C) 2010-2021, Christopher N. Hume.  All rights reserved.
//
//[2010-10-31 CNHume]Created Class
//
// Conditionals:
//
//#define DebugPeak

namespace Resource {
  using static Logging.Logger;

  using System;
  using System.Collections.Generic;
  using System.Diagnostics;

  class Pool<T> where T : class, new() {
    #region Constants
    const Int32 nDefaultAllocation = 12;
    #endregion

    #region Constructors
    public Pool(String Name = default) {
      this.Name = Name ?? typeof(T).Name;
      Inactive = new Stack<T>();
      Clear();
    }
    #endregion

    #region Methods
    public void Clear() {
      ActivePeak = ActiveCount = 0U;
    }

    private void Allocate(Int32 nAllocations) {
      for (var n = 0; n < nAllocations; n++)
        Inactive.Push(new T());         // Lazy<T> would defer invocation of initBoard()
    }

    public T Push() {
      if (Inactive is null)
        throw new ApplicationException("No Inactive Pool");

      if (Inactive.Count == 0)
        Allocate(nDefaultAllocation);

      var top = Inactive.Pop();
      Debug.Assert(top is not null, "Empty Inactive Pool");
      IncActive();
      return top;
    }

    public void Pop(ref T top) {
      if (top is null)
        throw new ArgumentNullException("Pool.pop() called with null argument");

      Inactive.Push(top);
      DecActive();
      top = default(T);
    }

    private void IncActive() {
      if (ActivePeak < ++ActiveCount) {
        ActivePeak = ActiveCount;
#if DebugPeak
        DisplayActive();
#endif
      }
    }

    private void DecActive() {
      ActiveCount--;
    }

    public void DisplayActive() {
      LogInfo(Level.data, $"{Name} Count = {ActiveCount}, Peak = {ActivePeak}");
    }
    #endregion

    #region Static Fields
    private Stack<T> Inactive;
    public UInt32 ActiveCount;
    public UInt32 ActivePeak;
    public String Name;
    #endregion
  }
}
